src/components/Features/Editor/Editor.js

```diff
@@ -0,0 +1,352 @@
+import {addJob} from '../../../api/sharp';
+import {addTransaction} from '../../../api/starknet';
+import {ProgramContext} from '../../../context/program/program-context';
+import {TabsContext} from '../../../context/tabs/tabs-context';
+import {promiseHandler} from '../../../utils/promise-handler';
+import {readFile} from '../../../utils/read-file';
+import {Box} from '../../Containers/Box/Box';
+import {DeployModal} from '../../UI/Modal/DeployModal';
+import {ProveModal} from '../../UI/Modal/ProveModal';
+import {DeployOutput} from '../Output/DeployOutput';
+import {ProveOutput} from '../Output/ProveOutput';
+import {CodeEditor} from './CodeEditor/CodeEditor';
+import styles from './Editor.module.scss';
+import {EditorButtons} from './EditorButtons/EditorButtons';
+
+const EDITOR_TITLE_TXT = 'editor';
+const COMPILING_MSG = 'Compiling...';
+const COMPILE_MSG = 'The contract was compiled successfully.';
+const SHARP_MSG = 'Sending job to SHARP...';
+const DEPLOYING_MSG = 'Deploying the contract...';
+const DEPLOY_FAILED_MSG = ['Got an error while deploying:'];
+const DEPLOY_ERROR_MSG = [
+  'The "%lang starknet" directive is missing.',
+  'Only StarkNet contracts can be deployed.',
+  'See the "StarkNet contract" lesson in the "All lessons" menu.'
+];
+
+const Editor = ({filePath}) => {
+  const program = useContext(ProgramContext);
+  const tabs = useContext(TabsContext);
+  const codeEditorRef = useRef(null);
+  const [marks, setMarks] = useState([]);
+  const [showProveModal, setShowProveModal] = useState(false);
+  const [showDeployModal, setShowDeployModal] = useState(false);
+
+  useEffect(() => {
+    async function loadFile() {
+      if (filePath) {
+        const content = await readFile(filePath);
+        program.setEditorContent(content);
+      } else {
+        program.setEditorContent('');
+      }
+    }
+
+    loadFile();
+  }, []);
+
+  useEffect(() => {
+    if (program.currentStep !== -1) {
+      clearMarks();
+      const entry = program.trace[program.currentStep];
+      const loc = program.instLocations[entry.pc];
+      addMark(loc);
+      scroll();
+    }
+  }, [program.currentStep]);
+
+  useEffect(() => {
+    if (!program.isDebugging) {
+      clearMarks();
+    }
+  }, [program.isDebugging]);
+
+  useEffect(() => {
+    updateLinks();
+  }, [program.content]);
+
+  useEffect(() => {
+    const {editor} = codeEditorRef.current;
+    if (program.isRunning) {
+      editor.setOption('readOnly', 'nocursor');
+    } else {
+      editor.setOption('readOnly', false);
+    }
+  }, [program.isRunning]);
+
+  const scroll = () => {
+    const target = document.getElementsByClassName(
+      'currentInstruction'
+    )[0];
+    if (target) {
+      const targetRect = target.getBoundingClientRect();
+      const scrollParent = codeEditorRef.current.ref.parentNode;
+      const scrollParentRect = scrollParent.getBoundingClientRect();
+      if (targetRect.bottom > scrollParentRect.bottom - 50) {
```

Comment:
where do these constants come from (50,64,32 below)?
wouldn't it be better to name them for what they mean?

Reply:
Will move them to constants.

Reply:
done